File: webApp/src/lib/chapters.test.ts
import { describe, expect, it } from 'vitest';
import {
  groupIntoChapters,
  formatOrdinal,
  chapterOf,
  localDateOf,
  type ChapterMedia,
} from './chapters';

let seq = 0;
function media(
  overrides: Partial<ChapterMedia> & { takenAt?: string | null },
): ChapterMedia {
  seq += 1;
  return {
    id: `m${seq}`,
    type: 'PHOTO',
    uploadedAt: '2026-04-14T10:00:00Z',
    ...overrides,
  } as ChapterMedia;
}

// Use local-noon timestamps so viewer-timezone grouping is deterministic
// regardless of the machine's offset.
const d = (day: number, hour = 12) =>
  new Date(2026, 3, day, hour, 0, 0).toISOString();

describe('groupIntoChapters — the five exact rules', () => {
  it('rule 1: sorts by takenAt ?? uploadedAt ascending and groups by local date', () => {
    const items = [
      media({ takenAt: d(12, 17) }),
      media({ takenAt: d(11, 9) }),
      media({ takenAt: d(11, 20) }),
    ];
    const chapters = groupIntoChapters(items);
    expect(chapters).toHaveLength(2);
    expect(chapters[0].items.map((m) => m.takenAt)).toEqual([d(11, 9), d(11, 20)]);
    expect(chapters[1].items[0].takenAt).toBe(d(12, 17));
  });

  it('rule 2: ordinals are ordinal within the trip — gaps do not create empty chapters', () => {
    const items = [
      media({ takenAt: d(11) }),
      media({ takenAt: d(15) }), // 3-day gap
      media({ takenAt: d(16) }),
    ];
    const chapters = groupIntoChapters(items);
    expect(chapters.map((c) => c.ordinal)).toEqual([1, 2, 3]);
  });

  it('rule 3: null takenAt groups by uploadedAt, no numeral, sorted after dated', () => {
    const items = [
      media({ takenAt: null, uploadedAt: d(10) }), // uploaded BEFORE the dated days
      media({ takenAt: d(12) }),
      media({ takenAt: d(11) }),
    ];
    const chapters = groupIntoChapters(items);
    expect(chapters).toHaveLength(3);
    // Dated chapters first, in date order, with ordinals.
    expect(chapters[0].dated).toBe(true);
    expect(chapters[0].ordinal).toBe(1);
    expect(chapters[1].ordinal).toBe(2);
    // Undated last, no ordinal, even though its date precedes the dated ones.
    expect(chapters[2].dated).toBe(false);
    expect(chapters[2].ordinal).toBeNull();
  });

  it('rule 4: single-date trip renders one chapter with the numeral suppressed', () => {
    const items = [media({ takenAt: d(11, 9) }), media({ takenAt: d(11, 15) })];
    const chapters = groupIntoChapters(items);
    expect(chapters).toHaveLength(1);
    expect(chapters[0].ordinal).toBeNull();
    expect(chapters[0].dated).toBe(true);
  });

  it('rule 4 counter-case: a single dated chapter plus an undated one keeps the numeral', () => {
    const items = [
      media({ takenAt: d(11) }),
      media({ takenAt: null, uploadedAt: d(14) }),
    ];
    const chapters = groupIntoChapters(items);
    expect(chapters[0].ordinal).toBe(1);
    expect(chapters[1].ordinal).toBeNull();
  });

  it('rule 5: the lead plate is the earliest item of the day', () => {
    const early = media({ takenAt: d(11, 6) });
    const late = media({ takenAt: d(11, 18) });
    const chapters = groupIntoChapters([late, early]);
    expect(chapters[0].lead.id).toBe(early.id);
    expect(chapters[0].rest.map((m) => m.id)).toEqual([late.id]);
  });

  it('counts photos and videos per chapter', () => {
    const items = [
      media({ takenAt: d(11) }),
      { ...media({ takenAt: d(11) }), type: 'VIDEO' as const },
    ];
    const [chapter] = groupIntoChapters(items);
    expect(chapter.photoCount).toBe(1);
    expect(chapter.videoCount).toBe(1);
  });

  it('chapterOf finds the chapter containing a media id', () => {
    const a = media({ takenAt: d(11) });
    const b = media({ takenAt: d(12) });
    const chapters = groupIntoChapters([a, b]);
    expect(chapterOf(chapters, b.id)?.ordinal).toBe(2);
    expect(chapterOf(chapters, 'nope')).toBeUndefined();
  });
});

describe('formatOrdinal', () => {
  it('pads to two digits', () => {
    expect(formatOrdinal(1)).toBe('01');
    expect(formatOrdinal(12)).toBe('12');
  });
  it('renders Arabic-Indic digits for Arabic display typography', () => {
    expect(formatOrdinal(1, true)).toBe('٠١');
    expect(formatOrdinal(23, true)).toBe('٢٣');
  });
});

describe('localDateOf', () => {
  it('returns the viewer-local calendar date', () => {
    const iso = new Date(2026, 3, 11, 23, 30).toISOString();
    expect(localDateOf(iso)).toBe('2026-04-11');
  });
});

File: shared/src/commonTest/kotlin/com/jameeli/thykra/chapters/ChaptersTest.kt
package com.jameeli.thykra.chapters

import com.jameeli.thykra.model.MediaType
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The same cases as the web's `chapters.test.ts`, so the two implementations stay one
 * grouping rather than two. All instants are UTC and the tests pin the timezone, so a CI
 * box in another zone gets the same answer.
 */
class ChaptersTest {

    private data class Media(
        override val id: String,
        override val type: MediaType = MediaType.PHOTO,
        override val takenAt: Instant? = null,
        override val uploadedAt: Instant,
        override val width: Int? = null,
        override val height: Int? = null,
    ) : ChapterMedia, HasDimensions

    private val utc = TimeZone.UTC

    private fun at(iso: String) = Instant.parse(iso)

    @Test
    fun `rule 1 - sorts by takenAt or uploadedAt and groups by local date`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("b", takenAt = at("2026-04-12T09:00:00Z"), uploadedAt = at("2026-04-14T00:00:00Z")),
                Media("a", takenAt = at("2026-04-11T18:00:00Z"), uploadedAt = at("2026-04-14T00:00:00Z")),
                Media("c", takenAt = at("2026-04-12T21:00:00Z"), uploadedAt = at("2026-04-14T00:00:00Z")),
            ),
            utc,
        )

        assertEquals(2, chapters.size)
        assertEquals(LocalDate(2026, 4, 11), chapters[0].date)
        assertEquals(listOf("a"), chapters[0].items.map { it.id })
        assertEquals(listOf("b", "c"), chapters[1].items.map { it.id })
    }

    @Test
    fun `rule 2 - ordinals are ordinal within the trip, gaps make no empty chapters`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("a", takenAt = at("2026-04-11T10:00:00Z"), uploadedAt = at("2026-04-11T10:00:00Z")),
                Media("b", takenAt = at("2026-04-15T10:00:00Z"), uploadedAt = at("2026-04-15T10:00:00Z")),
            ),
            utc,
        )

        assertEquals(listOf(1, 2), chapters.map { it.ordinal })
        assertEquals(listOf("d:2026-04-11", "d:2026-04-15"), chapters.map { it.key })
    }

    @Test
    fun `rule 3 - no takenAt groups by uploadedAt, no numeral, sorted after dated`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("undated", uploadedAt = at("2026-04-10T10:00:00Z")),
                Media("dated", takenAt = at("2026-04-12T10:00:00Z"), uploadedAt = at("2026-04-12T10:00:00Z")),
            ),
            utc,
        )

        assertEquals(2, chapters.size)
        assertTrue(chapters[0].dated)
        assertEquals(1, chapters[0].ordinal)
        // The undated chapter sorts last even though it was uploaded first.
        assertTrue(!chapters[1].dated)
        assertNull(chapters[1].ordinal)
        assertEquals("u:2026-04-10", chapters[1].key)
    }

    @Test
    fun `rule 4 - single-date trip renders one chapter with the numeral suppressed`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("a", takenAt = at("2026-04-12T08:00:00Z"), uploadedAt = at("2026-04-12T08:00:00Z")),
                Media("b", takenAt = at("2026-04-12T19:00:00Z"), uploadedAt = at("2026-04-12T19:00:00Z")),
            ),
            utc,
        )

        assertEquals(1, chapters.size)
        assertNull(chapters[0].ordinal)
    }

    @Test
    fun `rule 4 counter-case - one dated chapter plus an undated one keeps the numeral`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("a", takenAt = at("2026-04-12T08:00:00Z"), uploadedAt = at("2026-04-12T08:00:00Z")),
                Media("b", uploadedAt = at("2026-04-14T08:00:00Z")),
            ),
            utc,
        )

        assertEquals(2, chapters.size)
        assertEquals(1, chapters[0].ordinal)
    }

    @Test
    fun `rule 5 - the lead is the earliest item of the day`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("late", takenAt = at("2026-04-12T21:00:00Z"), uploadedAt = at("2026-04-12T21:00:00Z")),
                Media("early", takenAt = at("2026-04-12T06:00:00Z"), uploadedAt = at("2026-04-12T06:00:00Z")),
            ),
            utc,
        )

        assertEquals("early", chapters[0].lead.id)
        assertEquals(listOf("late"), chapters[0].rest.map { it.id })
    }

    @Test
    fun `counts photos and videos per chapter`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("p1", takenAt = at("2026-04-12T06:00:00Z"), uploadedAt = at("2026-04-12T06:00:00Z")),
                Media("v1", MediaType.VIDEO, at("2026-04-12T07:00:00Z"), at("2026-04-12T07:00:00Z")),
                Media("p2", takenAt = at("2026-04-12T08:00:00Z"), uploadedAt = at("2026-04-12T08:00:00Z")),
            ),
            utc,
        )

        assertEquals(2, chapters[0].photoCount)
        assertEquals(1, chapters[0].videoCount)
    }

    @Test
    fun `chapterOfMedia finds the chapter holding an id`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("a", takenAt = at("2026-04-11T10:00:00Z"), uploadedAt = at("2026-04-11T10:00:00Z")),
                Media("b", takenAt = at("2026-04-12T10:00:00Z"), uploadedAt = at("2026-04-12T10:00:00Z")),
            ),
            utc,
        )

        assertEquals("d:2026-04-12", chapters.chapterOfMedia("b")?.key)
        assertNull(chapters.chapterOfMedia("missing"))
    }

    @Test
    fun `heroCandidate is the largest-area photo, and never a video`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("small", takenAt = at("2026-04-12T06:00:00Z"), uploadedAt = at("2026-04-12T06:00:00Z"), width = 800, height = 600),
                Media("video", MediaType.VIDEO, at("2026-04-12T07:00:00Z"), at("2026-04-12T07:00:00Z"), width = 4000, height = 3000),
                Media("big", takenAt = at("2026-04-12T08:00:00Z"), uploadedAt = at("2026-04-12T08:00:00Z"), width = 4032, height = 3024),
            ),
            utc,
        )

        assertEquals("small", chapters[0].lead.id)
        assertEquals("big", chapters[0].heroCandidate.id)
    }

    @Test
    fun `heroCandidate falls back to the lead when nothing carries dimensions`() {
        val chapters = groupIntoChapters(
            listOf(
                Media("a", takenAt = at("2026-04-12T06:00:00Z"), uploadedAt = at("2026-04-12T06:00:00Z")),
                Media("b", takenAt = at("2026-04-12T08:00:00Z"), uploadedAt = at("2026-04-12T08:00:00Z")),
            ),
            utc,
        )

        assertEquals("a", chapters[0].heroCandidate.id)
    }

    @Test
    fun `formatOrdinal pads to two digits and can render Arabic-Indic`() {
        assertEquals("01", formatOrdinal(1))
        assertEquals("11", formatOrdinal(11))
        assertEquals("٠٢", formatOrdinal(2, arabicIndic = true))
        assertEquals("١٢", formatOrdinal(12, arabicIndic = true))
    }

    @Test
    fun `localDateOf returns the viewer-local calendar date`() {
        // 23:30 UTC on the 12th is the 13th in Amman (UTC+3).
        val instant = at("2026-04-12T23:30:00Z")
        assertEquals(LocalDate(2026, 4, 12), localDateOf(instant, TimeZone.UTC))
        assertEquals(LocalDate(2026, 4, 13), localDateOf(instant, TimeZone.of("Asia/Amman")))
    }
}
